### ISSUE TITLE
action distribution for estimating V

### ISSUE THREAD

**Issue by immars** [haarnoja/softqlearning#8]
Hi,
First of all, thanks for this inspiring work!

In 
https://github.com/haarnoja/softqlearning/blob/59c0bbb7d665616f796ab101de65227c89ffd318/softqlearning/algorithms/sql.py#L164

it seems to me that action is sampled from uniform distribution when estimating `V_{soft}` .

In Sec. 3.2. of your original paper, it is stated that
```
For q_a we have more options. A
convenient choice is a uniform distribution. However, this
choice can scale poorly to high dimensions. A better choice
is to use the current policy, which produces an unbiased
estimate of the soft value as can be confirmed by substi-
tution.
```

have you experimented with sampling from current policy to estimate `V`? Or, how good does uniform distribution do in practice, especially in higher dimensional cases?

thanks,

**Comment by immars** [comment]
ok, i see.
Thanks for the reply!

**Comment by Megultron** [comment]
I could be totally misunderstanding, but doesn't appendix C.2 talk about how one can use the sampling network for q_a' and derive the corresponding densities so long as the jacobian of a'/epsilon' is non-singular?

**Comment by haarnoja** [comment]
I see, that's indeed confusing. You are right in that we could compute the log probs if the sampling network is invertible. My feeling is that, in our case, the network does not remain invertible, and that the log probs we would obtain that way are wrong. We initially experimented with this trick (and that's why we discuss it in the appendix), but in the end, uniform samples worked better. We'll fix this in the next version of the paper, thanks for pointing it out!

**Comment by Megultron** [comment]
My pleasure! Glad I was sort of on the right track. That's very interesting, especially since non-singular weight matrices or choice of activation function are the only thing off the top of my head that might make a feedforward net non-invertible. I might play around with that.

**Comment by SJTUGuofei** [comment]
Hi, I have a few maybe stupid queations:
Also in softqlearning/softqlearning/algorithms/sql.py
1.  I don't understand why should do this "next_value += self._action_dim * np.log(2)" for computing the value?

2.  For computing $\hatQ$ in Eq. 11:
# \hat Q in Equation 11:
        ys = tf.stop_gradient(self._reward_scale * self._rewards_pl + (
            1 - self._terminals_pl) * self._discount * next_value)
It seems that only one sample used for computing the Expectation (below Eq. 11), does this reasonable , and is there accuracy guarantee for the obtained $\hat Q$?

Thank you!

**Comment by SJTUGuofei** [comment]
I see.
Thank you so much!

**Comment by zhanpenghe** [comment]
Why are we calculating the V with next observation? In the paper, it is Q_st_a'. Is it my misunderstanding?
